BigQuery Storage Write API: JsonStreamWriter cannot transform BigDecimal with 9+ decimals to BigNumeric

#### Environment details

1. OS type and version: OSX 11.6.3
2. Java version: OpenSDK 11
3.  version(s): v2.28.3 and later

#### Steps to reproduce

  1. Define a TableSchema with a field of type `BIGNUMERIC`
  2. Serialize a JSON message to Proto containing more than 9 decimals, eg. the number `24.6789012345`


#### Code example

This unit test could be added to `JsonToProtoMessageTest` to reproduce it:
```java
@Test
  public void testBigDecimalToBigNumericConversion() {
    TableSchema ts =
        TableSchema.newBuilder()
            .addFields(
                0,
                TableFieldSchema.newBuilder()
                    .setName("bignumeric")
                    .setType(TableFieldSchema.Type.BIGNUMERIC)
                    .setMode(TableFieldSchema.Mode.REPEATED)
                    .build())
            .build();
    TestBignumeric expectedProto =
        TestBignumeric.newBuilder()
            .addBignumeric(
                BigDecimalByteStringEncoder.encodeToBigNumericByteString(
                    new BigDecimal("24.6789012345")))
            .build();
    JSONObject json = new JSONObject();
    json.put("bignumeric", Collections.singletonList(new BigDecimal("24.6789012345")));
    DynamicMessage protoMsg =
        JsonToProtoMessage.convertJsonToProtoMessage(TestBignumeric.getDescriptor(), ts, json);
    assertEquals(expectedProto, protoMsg);
  }
```

#### Stack trace
```
[ERROR] com.google.cloud.bigquery.storage.v1.JsonToProtoMessageTest.testBigDecimalToBigNumericConversion  Time elapsed: 0.007 s  <<< ERROR!
com.google.cloud.bigquery.storage.v1.Exceptions$FieldParseError: ByteString scale cannot exceed 9: 24.6789012345
	at com.google.cloud.bigquery.storage.v1.JsonToProtoMessage.convertJsonToProtoMessageImpl(JsonToProtoMessage.java:247)
	at com.google.cloud.bigquery.storage.v1.JsonToProtoMessage.convertJsonToProtoMessage(JsonToProtoMessage.java:139)
	at com.google.cloud.bigquery.storage.v1.JsonToProtoMessageTest.testBigDecimalToBigNumericConversion(JsonToProtoMessageTest.java:1437)
...
```

#### Any additional information below

On the following lines, `encodeToBigNumericByteString` should be called instead of `encodeToNumericByteString`:

- https://github.com/googleapis/java-bigquerystorage/blob/v2.28.3/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessage.java#L358
- https://github.com/googleapis/java-bigquerystorage/blob/v2.28.3/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessage.java#L606

If you agree on this proposed approach, we can create PR with the following fix:
https://github.com/googleapis/java-bigquerystorage/compare/main...tradecloud:java-bigquerystorage:TC-8710-serialize-bigdecimal

--- a/README.md
+++ b/README.md
@@ -56,13 +56,13 @@ implementation 'com.google.cloud:google-cloud-bigquerystorage'
 If you are using Gradle without BOM, add this to your dependencies:
 
 ```Groovy
-implementation 'com.google.cloud:google-cloud-bigquerystorage:2.33.0'
+implementation 'com.google.cloud:google-cloud-bigquerystorage:2.33.1'
 ```
 
 If you are using SBT, add this to your dependencies:
 
 ```Scala
-libraryDependencies += "com.google.cloud" % "google-cloud-bigquerystorage" % "2.33.0"
+libraryDependencies += "com.google.cloud" % "google-cloud-bigquerystorage" % "2.33.1"
 ```
 
 ## Authentication

--- a/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessage.java
+++ b/google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessage.java
@@ -355,7 +355,7 @@ public class JsonToProtoMessage {
             } else if (val instanceof BigDecimal) {
               protoMsg.setField(
                   fieldDescriptor,
-                  BigDecimalByteStringEncoder.encodeToNumericByteString((BigDecimal) val));
+                  BigDecimalByteStringEncoder.encodeToBigNumericByteString((BigDecimal) val));
               return;
             }
           }
@@ -603,7 +603,7 @@ public class JsonToProtoMessage {
             } else if (val instanceof BigDecimal) {
               protoMsg.addRepeatedField(
                   fieldDescriptor,
-                  BigDecimalByteStringEncoder.encodeToNumericByteString((BigDecimal) val));
+                  BigDecimalByteStringEncoder.encodeToBigNumericByteString((BigDecimal) val));
               added = true;
             }
           }

--- a/google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessageTest.java
+++ b/google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1/JsonToProtoMessageTest.java
@@ -984,7 +984,7 @@ public class JsonToProtoMessageTest {
                 BigDecimalByteStringEncoder.encodeToNumericByteString(new BigDecimal(5D)))
             .setTestBignumeric(
                 BigDecimalByteStringEncoder.encodeToBigNumericByteString(
-                    new BigDecimal("578960446186580977117854925043439539266.3")))
+                    new BigDecimal("578960446186580977117854925043439539266.3222222222")))
             .addTestBignumericStr(
                 BigDecimalByteStringEncoder.encodeToBigNumericByteString(new BigDecimal("1.23")))
             .setTestBignumericShort(
@@ -1047,9 +1047,7 @@ public class JsonToProtoMessageTest {
     json.put("test_numeric_float", 4f);
     json.put("test_numeric_double", 5D);
     json.put(
-        "test_bignumeric",
-        BigDecimalByteStringEncoder.encodeToBigNumericByteString(
-            new BigDecimal("578960446186580977117854925043439539266.3")));
+        "test_bignumeric", new BigDecimal("578960446186580977117854925043439539266.3222222222"));
     json.put("test_bignumeric_str", new JSONArray(new String[] {"1.23"}));
     json.put("test_bignumeric_short", 1);
     json.put("test_bignumeric_int", 2);
@@ -1412,6 +1410,31 @@ public class JsonToProtoMessageTest {
     assertEquals(expectedProto, protoMsg);
     json.put("numeric", new Float(24.678));
     protoMsg = JsonToProtoMessage.convertJsonToProtoMessage(TestNumeric.getDescriptor(), ts, json);
+    assertEquals(expectedProto, protoMsg);
+  }
+
+  @Test
+  public void testBigDecimalToBigNumericConversion() {
+    TableSchema ts =
+        TableSchema.newBuilder()
+            .addFields(
+                0,
+                TableFieldSchema.newBuilder()
+                    .setName("bignumeric")
+                    .setType(TableFieldSchema.Type.BIGNUMERIC)
+                    .setMode(TableFieldSchema.Mode.REPEATED)
+                    .build())
+            .build();
+    TestBignumeric expectedProto =
+        TestBignumeric.newBuilder()
+            .addBignumeric(
+                BigDecimalByteStringEncoder.encodeToBigNumericByteString(
+                    new BigDecimal("24.6789012345")))
+            .build();
+    JSONObject json = new JSONObject();
+    json.put("bignumeric", Collections.singletonList(new BigDecimal("24.6789012345")));
+    DynamicMessage protoMsg =
+        JsonToProtoMessage.convertJsonToProtoMessage(TestBignumeric.getDescriptor(), ts, json);
     assertEquals(expectedProto, protoMsg);
   }
 

